src/template-pallet.md

```diff
@@ -0,0 +1,263 @@
+
+> **This pallet is meant only as an example.** We're including it to be helpful for future pallet authors
+that want to use the embedded native IPFS node to suit their needs.
+
+If you're familiar with Substrate and the Framework for Runtime Aggregation of Modularized Entities
+(FRAME), you can simply view the [source code] for this pallet. Otherwise, read on as we go through
+the code step by step.
+
+Please note the order in which these concepts are explained here is not necessarily the order that
+they appear in the code.
+
+You can also learn more by following the [Building a Custom Pallet] tutorial.
+
+[Building a Custom Pallet]: https://substrate.dev/docs/en/tutorials/build-a-dapp/pallet
+[source code]: https://github.com/rs-ipfs/substrate/blob/offchain_ipfs_bleeding_edge/bin/node-template/pallets/template/src/lib.rs
+[pallet]: https://substrate.dev/docs/en/knowledgebase/runtime/pallets
+
+## Prelude
+
+We start by using items from the native runtime. Our pallet is `no_std` since we're targeting `Wasm`
+
+```rust,ignore
+#![cfg_attr(not(feature = "std"), no_std)]
+// ...
+use sp_core::offchain::{
+  Duration, IpfsRequest, IpfsResponse, OpaqueMultiaddr, Timestamp
+};
+// ...
+use sp_runtime::offchain::ipfs;
+```
+
+## Command Types
+
+When your JSON-RPC calls are received by the pallet, the requests are
+expressed as `___Command` enums and stored in the off-chain worker storage as
+a queue, to be ingested by your native runtime and passed to IPFS.
+
+Derive attributes are omitted.
+
+```rust,ignore
+// Commands involved in peer-to-peer connections
+enum ConnectionCommand {
+    ConnectTo(OpaqueMultiaddr),
+    DisconnectFrom(OpaqueMultiaddr),
+}
+
+// Commands that add, remove, pin, unpin, and output data
+enum DataCommand {
+    AddBytes(Vec<u8>),
+    CatBytes(Vec<u8>),
+    InsertPin(Vec<u8>),
+    RemoveBlock(Vec<u8>),
+    RemovePin(Vec<u8>),
+}
+
+// Commands that query the distributed hash table (DHT)
+// for peers and content
+enum DhtCommand {
+    FindPeer(Vec<u8>),
+    GetProviders(Vec<u8>),
+}
+```
+
+## The runtime configuration trait
+
+The `system::Trait` trait (not to be confused with the Rust `trait` keyword), allows you to define
+which capabilities from the runtime you want to include, and how you want to use them. You can also
+"tightly couple" your pallet to other pallets by adding their `Trait`s to your pallet's
+inherited trait list.
+
+Here, however, we keep things simple by:
+
+1. Loosely coupling this pallet by leaving out inherited traits
+2. Including only the required `Event` type
+
+```rust,ignore
+/// The pallet's configuration trait.
+pub trait Trait: system::Trait { // Use traits here to tightly couple to runtime
+    /// The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+}
+```
+
+Later in the code, we implement some helper functions on this trait. The function bodies are
+omitted for brevity's sake.
```

Comment:
These helper functions are implemented for the Module struct instead of the trait Trait